### PR TITLE
drivers: bluetooth: hci: Introduce ST SPI protocol V2

### PR DESCRIPTION
**Pull request by HoZHel**
Update structure of spi.c to have a better and cleaner separation
between STMicroelectronics and Zephyr SPI protocol.

Update existing boards according to the new DTS approach.

Remove BT_SPI_BLUENRG as it is become redundant according to
the new changes.

Introduce STMicroelectronics SPI protocol V2 which is used in BlueNRG-1
and successor devices.

Add "st,hci-spi-v2.yaml" and use compatible property to enable
the protocol for the required boards.

Add BLE support for SensorTile.box board.

The new changes have been tested on 96b_carbon, disco_l475_iot1, b_l4s5i_iot01a,
stm32l562e_dk, x_nucleo_idb05a1, and sensortile_box boards.

### CODE REVIEW

File: drivers/bluetooth/hci/spi.c
Comment (avisconti) [inline drivers/bluetooth/hci/spi.c]: have you tested also the "else" side of the code?

File: boards/arm/sensortile_box/Kconfig.defconfig
Comment (avisconti) [inline boards/arm/sensortile_box/Kconfig.defconfig:11]: Why not ```depends on BT``` here?

Comment (erwango) [reply boards/arm/sensortile_box/Kconfig.defconfig:11]: `if` is a better option when you have several symbols impacted, as this is the case here.
`depends on` is well suited for single symbol dependency.
So current usage looks good to me.

File: drivers/bluetooth/hci/spi.c
Comment (HoZHel) [reply drivers/bluetooth/hci/spi.c]: Yes. It has been tested by 96b_carbon board which uses STM32F401RET and nRF51822 as main MCU and BLE module respectively.

File: drivers/bluetooth/hci/spi.c
Comment (avisconti) [inline drivers/bluetooth/hci/spi.c:183]: not sure I got why you added an arg and declaring it unused. Is it for future?

Comment (avisconti) [reply drivers/bluetooth/hci/spi.c:183]: ok, got it, sorry. It is because in v2 it is required...

File: drivers/bluetooth/hci/spi.c
Comment (avisconti) [inline drivers/bluetooth/hci/spi.c:227]: I wonder whether would be more readable to have all this routines as callbacks declared in separate files one for each protocol... Not sure it is worth doing it...

**Comment by avisconti** [comment]
It seems a very good an well written PR to me. Want to take some time to review it deeply before approving it though.

File: drivers/bluetooth/hci/spi.c
Comment (msmttchr) [reply drivers/bluetooth/hci/spi.c:227]: @avisconti, let me share my point of view about this.
Protocol have all the same structure, that is:
- header phase: used for command and data size definition
- data phase: where data is exchanged

What is more different is that header phase, while data phase is similar.
So, it could be a good compromise to keep everything together.
On the other hand, @JordanYates raised your point in the past as well.

Not a strong opinion from me, but I am little bit more inclined to keep the single file.

**Comment by JordanYates** [comment]
It would be nice to see a description of the differences between the two protocols (or a link to ST docs).
Does this affect two Zephyr devices using the built-in `samples/bluetooth/hci_spi` application?

File: drivers/bluetooth/hci/spi.c
Comment (avisconti) [reply drivers/bluetooth/hci/spi.c:227]: @msmttchr yes, sure. If changes are not that much, it might be useless to separate into multiple files.

**Comment by msmttchr** [comment]
> It would be nice to see a description of the differences between the two protocols (or a link to ST docs). Does this affect two Zephyr devices using the built-in `samples/bluetooth/hci_spi` application?

Regarding the protocol description, you can have a look at:
- V1: https://www.st.com/resource/en/user_manual/um1865-the-bluenrgms-bluetooth-le-stack-application-command-interface-aci-stmicroelectronics.pdf (section 5)
- V2: https://www.st.com/resource/en/datasheet/bluenrg-2n.pdf (section 6.2)

Regarding the protocol implemented in `samples/bluetooth/hci_spi`, this is not impacted in term of behavior, but it is impacted in terms of code restructuring.

**Comment by HoZHel** [comment]
The changes have been split into more commits to make it easier for review.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/bluetooth/hci/spi.c
Comment (HoZHel) [reply drivers/bluetooth/hci/spi.c:163]: Yes, since their use cases are different, and also `READ_CONDITION` is set to false for two protocols.

File: drivers/bluetooth/hci/spi.c
Comment (HoZHel) [reply drivers/bluetooth/hci/spi.c:262]: I didn't know about it. Thanks for informing.

Comment (HoZHel) [reply drivers/bluetooth/hci/spi.c]: It has been replaced by `WAIT_FOR`.

**Review by erwango — DISMISSED** [review]
LGTM for the parts I'm involved.

**Comment by jori-nordic** [comment]
@JordanYates you might want to review this, as it is a refactor of `spi.c`, not just a new driver.

**Comment by erwango** [comment]
Current twister issue should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/65540

**Review by avisconti — DISMISSED** [review]
LGTM

**Comment by HoZHel** [comment]
@jhedberg Would you please review the PR?

**Comment by carlescufi** [comment]
@JordanYates @jhedberg could you take a look please?

**Comment by jhedberg** [comment]
Looks like this will conflict with #65179. What's the plan to sync these two PRs? @arbrauns FYI.

**Comment by HoZHel** [comment]
> Looks like this will conflict with #65179. What's the plan to sync these two PRs? @arbrauns FYI.

Well, whichever is merged first, the other will rebase, resolve conflict, and force-push.

**Comment by jhedberg** [comment]
Hmm, GitHub doesn't indicate any conflicts, so maybe a rebase isn't even necessary. Would be good to double-check that nothing is broken with the combination of the two PRs, though.

File: samples/bluetooth/hci_usb_h4/sample.yaml
Comment (alwa-nordic) [inline samples/bluetooth/hci_usb_h4/sample.yaml:13]: You are adding onto a 'FIXME'. Why is this needed? Can it be resolved properly in some way?

**Review by alwa-nordic — DISMISSED** [review]

File: samples/bluetooth/hci_usb_h4/sample.yaml
Comment (HoZHel) [reply samples/bluetooth/hci_usb_h4/sample.yaml:13]: This test is related to Bluetooth over USB, and it is performed by Twister automatically when the board has "usb_device" and "ble" capability in its yaml file. However, this dependency for the test is not true that it indicates every board which has aforementioned capabilities should support Bluetooth over USB. At least, at the moment SensorTile.box does not support this feature. In addition, the comment "FIXME" is wrong as there is no problem to be fixed.

I may skip this test without using exclusion since there is already a misspelling in the board's yaml file regarding "usb_device" parameter, and intentionally or by mistake instead of "_" there is a white space.
Please, let me know which approach you prefer.

**Comment by HoZHel** [comment]
> Hmm, GitHub doesn't indicate any conflicts, so maybe a rebase isn't even necessary. Would be good to double-check that nothing is broken with the combination of the two PRs, though.

Even though we changed the same file, we didn't touch the same parts. So, generally there shouldn't be any conflict; however, I'll rebase and force push to double-check.

File: drivers/bluetooth/hci/spi.c
Comment (JordanYates) [inline drivers/bluetooth/hci/spi.c:511]: Writing partial data is worse than writing no data at all IMO.
This check can also never fail for the non-ST protocol, not sure whether the compiler is smart enough to figure that out though.

**Review by JordanYates — CHANGES_REQUESTED** [review]
Minor problem with writing partial packets.

Not part of the requested changes, but this file now contains enough `#ifdef` layers that I find it annoying to figure out which parts apply when. I think it would be worthwhile trying to split the various categories into dedicated files soon.

File: drivers/bluetooth/hci/spi.c
Comment (HoZHel) [reply drivers/bluetooth/hci/spi.c]: Done.

**Review by avisconti — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]
Approved on the parts I'm involved

**Review by erwango — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by msmttchr — APPROVED** [review]
LGTM. Thanks.

**Comment by msmttchr** [comment]
@jhedberg, unfortunately this PR is quite urgent for me, as it can unblock a certain number of boards to be supported in Zephyr.
Apparently, all the comments have been addressed, do you see any blocking point for merging it?

**Comment by jhedberg** [comment]
> @jhedberg, unfortunately this PR is quite urgent for me, as it can unblock a certain number of boards to be supported in Zephyr. Apparently, all the comments have been addressed, do you see any blocking point for merging it?

No, I don't see any showstoppers for merging. Not sure why it hasn't happened yet, since typically people from the release engineering team (with merge privileges) periodically check for PRs that are mergeable and merge them. @erwango I see you requested @fabiobaltieri for review - is there something still unclear or that was just to expedite getting this merged (I also have merge privileges, but wanted to check before I press "merge").

**Comment by jhedberg** [comment]
> No, I don't see any showstoppers for merging. Not sure why it hasn't happened yet, since typically people from the release engineering team (with merge privileges) periodically check for PRs that are mergeable and merge them. 

Answering myself:
"1 change requested

@JordanYates requested changes"

That's the reason the PR didn't show up in any "mergeable PR" filters.

**Comment by erwango** [comment]
> @erwango I see you requested @fabiobaltieri for review - is there something still unclear or that was just to expedite getting this merged (I also have merge privileges, but wanted to check before I press "merge").

The reason I added Fabio is that I though that maybe you were prevented to merge as you also are the assignee.

**Comment by fabiobaltieri** [comment]
> The reason I added Fabio is that I though that maybe you were prevented to merge as you also are the assignee.

Not a thing, the only limitation on individuals is something like: no author and reviewer and merger from the same company.

@JordanYates is indeed blocking though, can you make another pass?

**Review by JordanYates — APPROVED** [review]

**Pull request merged**
